python-squid/src/main/java/org/sonar/python/semantic/SymbolTableBuilderVisitor.java

```diff
@@ -0,0 +1,345 @@
+/*
+ * SonarQube Python Plugin
+ * Copyright (C) 2011-2017 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.python.semantic;
+
+import com.google.common.collect.ImmutableSet;
+import com.sonar.sslr.api.AstNode;
+import com.sonar.sslr.api.AstNodeType;
+import java.util.Collections;
+import java.util.Deque;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.LinkedList;
+import java.util.Map;
+import java.util.Set;
+import javax.annotation.CheckForNull;
+import javax.annotation.Nullable;
+import org.sonar.python.PythonVisitor;
+import org.sonar.python.PythonVisitorContext;
+import org.sonar.python.api.PythonGrammar;
+import org.sonar.python.api.PythonPunctuator;
+import org.sonar.sslr.ast.AstSelect;
+
+public class SymbolTableBuilderVisitor extends PythonVisitor {
+
+  private Map<AstNode, Scope> scopesByRootTree;
+  private Set<AstNode> allReadUsages;
+
+  public SymbolTable symbolTable() {
+    return new SymbolTablImpl(scopesByRootTree);
+  }
+
+  @Override
+  public void scanFile(PythonVisitorContext context) {
+    super.scanFile(context);
+    new FirstPhaseVisitor().scanFile(context);
+    new SecondPhaseVisitor().scanFile(context);
+  }
+
+  @Override
+  public void visitFile(AstNode node) {
+    scopesByRootTree = new HashMap<>();
+    allReadUsages = new HashSet<>();
+  }
+
+  private class ScopeVisitor extends PythonVisitor {
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Make this a "static" inner class. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2694)